src/ophyd_async/core/_utils.py

```diff
@@ -119,19 +123,38 @@ async def wait_for_connection(**coros: Awaitable[None]):
         raise NotConnected(exceptions)
 
 
-def get_dtype(typ: type) -> np.dtype | None:
+def get_dtype(datatype: type) -> np.dtype:
     """Get the runtime dtype from a numpy ndarray type annotation
 
     >>> import numpy.typing as npt
     >>> import numpy as np
     >>> get_dtype(npt.NDArray[np.int8])
     dtype('int8')
     """
-    if getattr(typ, "__origin__", None) == np.ndarray:
-        # datatype = numpy.ndarray[typing.Any, numpy.dtype[numpy.float64]]
-        # so extract numpy.float64 from it
-        return np.dtype(typ.__args__[1].__args__[0])  # type: ignore
-    return None
+    if not get_origin(datatype) == np.ndarray:
+        raise TypeError(f"Expected np.ndarray, got {datatype}")
+    # datatype = numpy.ndarray[typing.Any, numpy.dtype[numpy.float64]]
+    # so extract numpy.float64 from it
+    return np.dtype(get_args(get_args(datatype)[1])[0])
```

Comment:
Do we allow for np arrays without subtype? If so we need:

```suggestion
    if get_origin(datatype) == np.ndarray:
        # np.ndarray[typing.Any, np.dtype[np.float64]] returns np.float64
        return np.dtype(get_args(get_args(datatype)[1])[0])
    elif datatype == np.ndarray:
        return np.float64
    else:
        raise TypeError(f"Expected np.ndarray, got {datatype}")
```

If not then perhaps a modification to the error message that the numpy array must be explicitly typed.

Reply:
Updated the error message